File: solutions/day15_part1.py
with open("./input.txt", "r") as inp:
  wli: str = inp.read().rstrip()
  inp.close()

warehouse, movements = wli.split('\n\n')
warehouse = [list(row) for row in warehouse.split('\n')]
movements = list(movements.replace('\n', ''))
dxdy = {
    '^': (0, -1),
    '>': (1, 0),
    'v': (0, 1),
    '<': (-1, 0)
}

def checkWarehouse(map: list[list[str]]) -> None:
    for i in range(len(map)):
        for j in range(len(map[i])):
            print(map[i][j], end='')
        print()
    print()

def locateRobot(map: list[list[str]]) -> tuple[int, int]:
    for i in range(len(map)):
        for j in range(len(map[i])):
            if map[i][j] == '@':
                return (j, i)
    # return (-1, -1)

def moveRobot(map: list[list[str]], movements: list[str]) -> None:
    x, y = locateRobot(map)
    for i in range(len(movements)):
        movable = True
        dx, dy = dxdy[movements[i]]
        if dx != 0:
            lastBox = x
            while map[y][lastBox + dx] == 'O':
                lastBox += dx
            if map[y][lastBox + dx] == '.':
                map[y][x + dx], map[y][lastBox + dx] = map[y][lastBox + dx], map[y][x + dx]
            elif map[y][lastBox + dx] == '#':
                movable = False
        else:
            lastBox = y
            while map[lastBox + dy][x] == 'O':
                lastBox += dy
            if map[lastBox + dy][x] == '.':
                map[y + dy][x], map[lastBox + dy][x] = map[lastBox + dy][x], map[y + dy][x]
            elif map[lastBox + dy][x] == '#':
                movable = False
        if movable:
            map[y][x], map[y + dy][x + dx] = map[y + dy][x + dx], map[y][x]
            x += dx
            y += dy
    # print(i, dy, dx, lastBox, movable)
    # checkWarehouse(map)

# checkWarehouse(map)
map = [row[:] for row in warehouse]
moveRobot(map, movements)
coordsSum = 0
for i in range(1, len(map) - 1):
    for j in range(1, len(map[i]) - 1):
        if map[i][j] == 'O':
            coordsSum += i * 100 + j
print(coordsSum)

# if movements[i] == '^' and (warehouse[y - 1][x] == '.' or (warehouse[y - 1][x] == 'O' and warehouse[y - 2][x] != '#')):
#     if warehouse[y - 1][x] == 'O':
#         warehouse[y - 1][x], warehouse[y - 2][x] = warehouse[y - 2][x], warehouse[y - 1][x]
#     warehouse[y][x], warehouse[y - 1][x] = warehouse[y - 1][x], warehouse[y][x]
#     y -= 1
# elif movements[i] == 'v' and (warehouse[y + 1][x] == '.' or (warehouse[y + 1][x] == 'O' and warehouse[y + 2][x] != '#')):
#     if warehouse[y + 1][x] == 'O':
#         warehouse[y + 1][x], warehouse[y + 2][x] = warehouse[y + 2][x], warehouse[y + 1][x]
#     warehouse[y][x], warehouse[y + 1][x] = warehouse[y + 1][x], warehouse[y][x]
#     y += 1
# elif movements[i] == '<' and (warehouse[y][x - 1] == '.' or (warehouse[y][x - 1] == 'O' and warehouse[y][x - 2] != '#')):
#     if warehouse[y][x - 1] == 'O':
#         warehouse[y][x - 1], warehouse[y][x - 2] = warehouse[y][x - 2], warehouse[y][x - 1]
#     warehouse[y][x], warehouse[y][x - 1] = warehouse[y][x - 1], warehouse[y][x]
#     x -= 1
# elif movements[i] == '>' and (warehouse[y][x + 1] == '.' or (warehouse[y][x + 1] == 'O' and warehouse[y][x + 2] != '#')):
#     if warehouse[y][x + 1] == 'O':
#         warehouse[y][x + 1], warehouse[y][x + 2] = warehouse[y][x + 2], warehouse[y][x + 1]
#     warehouse[y][x], warehouse[y][x + 1] = warehouse[y][x + 1], warehouse[y][x]
#     x += 1

File: solutions/day15_part1.ts
import fs from "node:fs";

const raw = fs.readFileSync("./input.txt", "utf-8");
const rawData = raw.split("\n\n");
const warehouse = rawData[0].split("\n").map((row) => row.split(""));
const movements = rawData[1].replaceAll("\n", "").split("");
const directionsMap = new Map([
  ["^", [0, -1]],
  [">", [1, 0]],
  ["v", [0, 1]],
  ["<", [-1, 0]],
]);

// function displayWarehouse(map: string[][]): void {
//   console.log(map.map((row) => row.join("")).join("\n"));
// }

function locateStart(map: string[][]): [number, number] {
  for (let i = 0; i < map.length; i++) {
    for (let j = 0; j < map[i].length; j++) {
      if (map[i][j] === "@") return [j, i];
    }
  }
  return [-1, -1];
}

function moveRobot(map: string[][], movements: string[]): void {
  let [x, y] = locateStart(map);
  for (const movement of movements) {
    let movable = true;
    const [dx, dy] = directionsMap.get(movement)!;
    if (dx !== 0) {
      let lastBox = x;
      while (map[y][lastBox + dx] === "O") {
        lastBox += dx;
      }
      if (map[y][lastBox + dx] === ".") {
        const temp = map[y][x + dx];
        map[y][x + dx] = map[y][lastBox + dx];
        map[y][lastBox + dx] = temp;
      } else if (map[y][lastBox + dx] === "#") {
        movable = false;
      }
    } else {
      let lastBox = y;
      while (map[lastBox + dy][x] === "O") {
        lastBox += dy;
      }
      if (map[lastBox + dy][x] === ".") {
        const temp = map[y + dy][x];
        map[y + dy][x] = map[lastBox + dy][x];
        map[lastBox + dy][x] = temp;
      } else if (map[lastBox + dy][x] === "#") {
        movable = false;
      }
    }
    if (movable) {
      const temp = map[y][x];
      map[y][x] = map[y + dy][x + dx];
      map[y + dy][x + dx] = temp;
      x += dx;
      y += dy;
    }
  }
}

const newWarehouse: string[][] = JSON.parse(JSON.stringify(warehouse));
moveRobot(newWarehouse, movements);
const coordsSum = newWarehouse.reduce(
  (acc, row, i) =>
    acc +
    row.reduce(
      (accRow, position, j) =>
        position === "O" ? accRow + i * 100 + j : accRow,
      0
    ),
  0
);
console.log(coordsSum);
